In [104]:
import math

In [105]:
YL = 30
lambda_r = 3
lambda_p = 10
p = 0.4
q = 0.05
S = YL*4*10


In [163]:
Pr = [[0 for col in range(S)] for row in range(S)]

In [164]:
state_mapper = {}
state_mapper.setdefault(tuple([0, 1, 10]), 0)
state_mapper.setdefault(tuple([-2, -2, -2]), 1)
state_mapper.setdefault(tuple([-3, -3, -3]), 2)
visited_r = set();

In [160]:
def dfs_r(state): # use tuple as input 
    visited_r.add(state) # これが重要
    print "start dfs %s" %(state, )
    index = state_mapper.get( state )
    if index == 1:
        print "Lose state"
        return
    elif index == 2:
        print "Win state" 
        return
    else:
        # Not a terminal state, expand it
        Neighbors = generate_neighbors_r(state) # a set of tuples
        for Nei in Neighbors:
            Pr[index][state_mapper.get(Nei)] = 1
            if not Nei in visited_r:
                dfs_r(Nei)
    print "End dfs %s" %(state, )
    return

In [159]:
def generate_neighbors_r(state):
    yl = state[0]
    dw = state[1]
    tg = state[2]
    Neighbors = set()
    Neighbors.add( tuple( [-3, -3, -3] ) )
    if dw == 4:
        Neighbors.add( tuple( [-2, -2, -2] ) )
        for yl_ in range(yl, YL):
            if tg - (yl_ - yl) <= 0:
                state_ = tuple([yl_, 1, 10])
                add_mapper(state_)
                Neighbors.add(state_)
    else:
        for yl_ in range(yl, YL):
            if tg-(yl_ - yl) >= 1:
                state_ = tuple([yl_, dw+1, tg-(yl_ - yl)])
            else:
                state_ = tuple([yl_, 1, 10])
            add_mapper(state_)
            Neighbors.add(state_)
    return Neighbors

In [110]:
def add_mapper(state):
    if not state in state_mapper:
        state_mapper.setdefault(state, len(state_mapper.keys()))
        # print len(state_mapper.keys())

In [111]:
def generate_p_for_Pr():
    for state in state_mapper.keys():
        for state_ in state_mapper.keys():
            if Pr[state_mapper.get(state)][state_mapper.get(state_)] != 0:
                Pr[state_mapper.get(state)][state_mapper.get(state_)] = computePr(state, state_)
    

In [112]:
def computePr(state, state_):
    yl = state[0]
    dw = state[1]
    tg = state[2]
    yl_ = state_[0]
    dw_ = state_[1]
    tg_ = state_[2]
    if state_mapper.get(state_) == 1:
        # compute probability of lose
        pr = 0
        for i in range(tg):
            if yl+i < YL:
                pr = pr + poi(i, lambda_r)
    elif state_mapper.get(state_) == 2:
        # probability of win
        pr = 1
        for i in range(YL-yl):
            pr = pr - poi(i, lambda_r)
    else:
        # probability from state to state_
        try:
            pr = poi(yl_ - yl, lambda_r)
        except:
            print "ERROR: Start at %s" %(state, )
            print "ERROR: End at %s" %(state_, )
    return pr

In [10]:
def poi(k, lambda_):
    return lambda_**k * math.exp(-lambda_)/math.factorial(k)

In [165]:
dfs_r(tuple([0, 1, 10]))

start dfs (0, 1, 10)
start dfs (16, 1, 10)
start dfs (22, 2, 4)
start dfs (24, 3, 2)
start dfs (29, 1, 10)
start dfs (29, 2, 10)
start dfs (29, 3, 10)
start dfs (29, 4, 10)
start dfs (-2, -2, -2)
Lose state
start dfs (-3, -3, -3)
Win state
End dfs (29, 4, 10)
End dfs (29, 3, 10)
End dfs (29, 2, 10)
End dfs (29, 1, 10)
start dfs (24, 4, 2)
start dfs (28, 1, 10)
start dfs (28, 2, 10)
start dfs (29, 3, 9)
start dfs (29, 4, 9)
End dfs (29, 4, 9)
End dfs (29, 3, 9)
start dfs (28, 3, 10)
start dfs (28, 4, 10)
End dfs (28, 4, 10)
End dfs (28, 3, 10)
End dfs (28, 2, 10)
start dfs (29, 2, 9)
End dfs (29, 2, 9)
End dfs (28, 1, 10)
start dfs (26, 1, 10)
start dfs (26, 2, 10)
start dfs (28, 3, 8)
start dfs (28, 4, 8)
End dfs (28, 4, 8)
start dfs (29, 4, 7)
End dfs (29, 4, 7)
End dfs (28, 3, 8)
start dfs (27, 3, 9)
start dfs (27, 4, 9)
End dfs (27, 4, 9)
End dfs (27, 3, 9)
start dfs (26, 3, 10)
start dfs (26, 4, 10)
End dfs (26, 4, 10)
End dfs (26, 3, 10)
start dfs (29, 3, 7)
End dfs (29, 3, 7)
End

In [169]:
generate_p_for_Pr()

In [166]:
#state_mapper.get(tuple([17,4,9]))

92

In [170]:
#Pr[92][0:10]


[0, 0.9961970079383241, 1.6149048555909162e-05, 0, 0, 0, 0, 0, 0, 0]

In [171]:
# construct matrices for pass strategy
Pp = [[0 for col in range(S)] for row in range(S)]
visited_p = set();

In [162]:
def dfs_p(state): # use tuple as input 
    visited_p.add(state) # これが重要
    print "start dfs %s" %(state, )
    index = state_mapper.get( state )
    if index == 1:
        print "Lose state"
        return
    elif index == 2:
        print "Win state" 
        return
    else:
        # Not a terminal state, expand it
        Neighbors = generate_neighbors_p(state) # a set of tuples
        for Nei in Neighbors:
            Pp[index][state_mapper.get(Nei)] = computePp(state, Nei)
            if not Nei in visited_p:
                dfs_p(Nei)
    print "End dfs %s" %(state, )
    return

In [161]:
def generate_neighbors_p(state):
    yl = state[0]
    dw = state[1]
    tg = state[2]
    Neighbors = set()
    Neighbors.add( tuple( [-3, -3, -3] ) )
    Neighbors.add( tuple( [-2, -2, -2] ) )
    if dw == 4:
        for yl_ in range(yl, YL):
            if tg - (yl_ - yl) <= 0:
                state_ = tuple([yl_, 1, 10])
                add_mapper(state_)
                Neighbors.add(state_)
    else:
        for yl_ in range(yl, YL):
            if tg-(yl_ - yl) >= 1:
                state_ = tuple([yl_, dw+1, tg-(yl_ - yl)])
            else:
                state_ = tuple([yl_, 1, 10])
            add_mapper(state_)
            Neighbors.add(state_)
    return Neighbors

In [120]:
def computePp(state, state_):
    yl = state[0]
    dw = state[1]
    tg = state[2]
    yl_ = state_[0]
    dw_ = state_[1]
    tg_ = state_[2]
    ppc = 1-p-q
    if state_mapper.get(state_) == 1:
        # compute probability of lose
        pp = q
        if dw == 4:
            pp = pp + p
            for i in range(tg):
                if yl+i < YL:
                    pp = pp + ppc * poi(i, lambda_p)
    elif state_mapper.get(state_) == 2:
        # probability of win
        pp = ppc
        for i in range(YL-yl):
            pp = pp - ppc * poi(i, lambda_p)
    else:
        # probability from state to state_
        try:
            if yl_==yl:
                pp = ppc * poi(yl_ - yl, lambda_p) + p
            else:
                pp = ppc * poi(yl_ - yl, lambda_p)
        except:
            print "ERROR: Start at %s" %(state, )
            print "ERROR: End at %s" %(state_, )
    return pp

In [172]:
dfs_p(tuple([0, 1, 10]))

start dfs (0, 1, 10)
start dfs (16, 1, 10)
start dfs (22, 2, 4)
start dfs (24, 3, 2)
start dfs (29, 1, 10)
start dfs (29, 2, 10)
start dfs (29, 3, 10)
start dfs (29, 4, 10)
start dfs (-2, -2, -2)
Lose state
start dfs (-3, -3, -3)
Win state
End dfs (29, 4, 10)
End dfs (29, 3, 10)
End dfs (29, 2, 10)
End dfs (29, 1, 10)
start dfs (24, 4, 2)
start dfs (28, 1, 10)
start dfs (28, 2, 10)
start dfs (29, 3, 9)
start dfs (29, 4, 9)
End dfs (29, 4, 9)
End dfs (29, 3, 9)
start dfs (28, 3, 10)
start dfs (28, 4, 10)
End dfs (28, 4, 10)
End dfs (28, 3, 10)
End dfs (28, 2, 10)
start dfs (29, 2, 9)
End dfs (29, 2, 9)
End dfs (28, 1, 10)
start dfs (26, 1, 10)
start dfs (26, 2, 10)
start dfs (27, 3, 9)
start dfs (29, 4, 7)
End dfs (29, 4, 7)
start dfs (28, 4, 8)
End dfs (28, 4, 8)
start dfs (27, 4, 9)
End dfs (27, 4, 9)
End dfs (27, 3, 9)
start dfs (26, 3, 10)
start dfs (26, 4, 10)
End dfs (26, 4, 10)
End dfs (26, 3, 10)
start dfs (28, 3, 8)
End dfs (28, 3, 8)
start dfs (29, 3, 7)
End dfs (29, 3, 7)
End

In [173]:
#state_mapper.get(tuple([16,4,10]))

91

In [174]:
#Pp[92][0:10]

[0, 0.6330508233128953, 0.11464393798281904, 0, 0, 0, 0, 0, 0, 0]

In [175]:
N = len(state_mapper)
print N

518


In [176]:
# Terminal state, probabilities associated with it, and restrict Pr->Prr, Pp->Ppp
state_mapper.setdefault(tuple([-4, -4, -4]), N)
Pr[1][N] = 1
Pr[2][N] = 1
Pr[N][N] = 1
Pp[1][N] = 1
Pp[2][N] = 1
Pp[N][N] = 1

Prr = [[0 for col in range(N+1)] for row in range(N+1)]
for i in range(N+1):
    for j in range(N+1):
        Prr[i][j] = Pr[i][j]
        
Ppp = [[0 for col in range(N+1)] for row in range(N+1)]
for i in range(N+1):
    for j in range(N+1):
        Ppp[i][j] = Pp[i][j]
        
# construct payoff matrix
g = [[[0 for col in range(N+1)] for control in range(2)] for row in range(N+1)]

g[2][0][N] = 1
g[2][1][N] = 1

# initial payoffs 
g0 = [0 for col in range(N+1)] 

## Result till now
Now we have 
- Prr and Ppp as markov transition probabilities
- g as payoffs
- g0 as initial payoff
- a state_mapper to check state |-> int in 0~N, where N=518

In [126]:
import csv
# write the results as files
def map_writer(map, filename):
    f = open(filename, 'w')
    wr = csv.writer(f, delimiter=' ', quoting = csv.QUOTE_NONE)
    for key in map.keys():
        wr.writerow((map.get(key), key[0],key[1],key[2] ))
    f.close()

In [177]:
map_writer(state_mapper, 'state_mapper.txt')

In [178]:
import numpy as np
np.savetxt('Pr.txt', Prr, delimiter=' ')

In [179]:
np.savetxt('Pp.txt', Ppp, delimiter=' ')